### ISSUE TITLE
Visualizing 5mC/5hmC at only CG or CH sequence.

### ISSUE THREAD

**Issue by miwi610** [igvteam/igv#1479]
Dorado basecaller for ONT can call 5mC/5hmC not only at CpG, but also at CH sequence, now.
And many researchers, I think, like to visualize the modification at CG and CH separately, as IGV have those options at bisulfide mode.
If there are not so yet, please implement to visualize 5mC/5hmC modification of ONT bam at CG and CH sequence, separately, similar to bisulfite mode. I checked the newest 2.17.1 and could not find such option so far.
I am not sure if there are needs for 6mA modificaiton at only AT and A[C|G|A], too.

**Comment by jrobinso** [comment]
Basemod visulaization does not depend on context (CG or CH).   I'm not really sure what you are asking for,  IGV is going to display base mods chosen wherever they appear.

**Comment by jrobinso** [comment]
If you are asking to restrict base mods to sites with specific reference sequences that is not going to be implemented in IGV.

**Comment by miwi610** [comment]
There are options in current IGV at "Color alignments by" -> "bisulfite mode" -> "CG" for whole-genome bisulfite sequencing data.
I was asking if it is possible to do so for ONT data. Visualization of 5mC for CG sequence only, etc.
Is it possible?

**Comment by jrobinso** [comment]
Yes, if base modifications are recorded in the BAM file (MM / ML tags).  That has been available for over a year.  See the screenshot below, and look at the user guide,  perhaps I'm not understanding what you are asking for.
<img width="503" alt="Screen Shot 2024-01-29 at 10 36 37 PM" src="https://github.com/igvteam/igv/assets/933148/d2bb023d-66d0-4b14-9362-696683e18613">

**Comment by jrobinso** [comment]
https://igv.org/doc/desktop/#UserGuide/tracks/alignments/base_modifications/

**Comment by miwi610** [comment]
Sorry for my language problem.
Previous ONT data only contained 5mC/5hmC modification at CG sequences ("5mCG_5hmCG" option in dorado/megalodon), which was nicely visualized in IGV as shown in your screenshot and the UserGuide.
New dorado output has an option, "5mC_5hmC", to basecall 5mC/5hmC not only CG, but also CH, meaning all C and G (complement of C) have values and visualized in IGV.
In most of mammalian cells, most of CpG are methylated and CpH are unmetlylated, therefore in IGV, most of C and G (complement of C) are visualized as unmethylated and small number of C are methylated (CpG)(screenshot).
For instance, the IGV view shown in zoomed CGI region show methylated at CG and unmethylated at other Cs. If I zoom out, the region looks not much methylated, although all CpG in this region are fully methylated.
To reproduce what I used to see, visualizing only CG sequence, I can re-basecall with "5mCG_5hmCG" option for all my data. But if IGV can extract and visualize only CG modification, it is very useful and computer-resource frienly.
I hope you understand what I mean...
![スクリーンショット 2024-01-30 16 51 24](https://github.com/igvteam/igv/assets/139537326/5fd995da-30f5-4685-9980-2562c0658a03)
![スクリーンショット 2024-01-30 16 52 33](https://github.com/igvteam/igv/assets/139537326/372f800f-ff0c-4eb4-935d-6fb5e6d84404)

**Comment by jrobinso** [comment]
OK, I understand I think.  Your language is fine, I think the title of the ticket summarizes it well.  

To generalize this,  If I understand correctly, you are requesting that base modifications be shown only at specific contexts.  By that I mean only at genomic locations that contain a specific sequence.   I can see how that might be useful but it would be complex to implement, and I cannot see this being prioritized anytime soon.

**Comment by miwi610** [comment]
Yes! That is exactly what I meant.
I can imagine this is complexed step, especially for ONT bam data.
At least it is troubling me to decode MM and ML values with reference sequence and long CIGAR string, and find out values for what I need to know.
Thanks a lot for considering to include in the future version of IGV.

**Comment by jrobinso** [comment]
Let's leave it open for now, I'm just trying to be realistic about the effort and time.   @marcus1487 @ctsa do you have any thoughts on this?

**Comment by jrobinso** [comment]
@miwi610 Have you tried the single color option `base modification (all)`?  If you are just looking for methylated bases this might work better than the 2 color option for this situation.

**Comment by miwi610** [comment]
Yes. I noticed I was viewing with `IGV 2.16.2` with "`Color alignment by -> base modification (5mC)`" in the screenshot above.
I made 4 screenshots, attached.
For all of them, upper panel is viewing bam from `sup,5mC_5hmC,6mA` option in `dorado 0.5.2`, lower panel is that from `sup,5mCG_5hmCG` option (MM and ML at only CG target) of same ONT pod5 data.
What I am asking is to **visualize data in upper panel looks similar to lower panel.**
Blue/red colored pictures are shown in `IGV 2.17.1 `with `Color alignment by -> base modification (5mC)`.
Cyan colored pictures are shown with `Color alignment by -> base modification (5hmC)`.
1st and 3rd picture with region of interest marks at CGI near the gene promoter and non-CGI are further zoomed at ROI, 2nd and 4th pictures.
I hope you understand what I mean...
Because the modification levels (both 5mC/5hmC) at CH are low, zoomed view seems to look similar to what it is suppose to look (**low 5mC at CGI, high at non-CGI, higher 5hmC at CGI, lower at non-CGI**).
Therefore it is not-so-bad level in this scale, I just need to separately mark where the GC sequences are.
But if I zoom out to see wider area, the difference is masked and all area looks highly modified.
![スクリーンショット 2024-02-01 14 33 56](https://github.com/igvteam/igv/assets/139537326/5735bf02-0916-4fa6-9b4e-c4d39ed34c00)
![スクリーンショット 2024-02-01 14 32 49](https://github.com/igvteam/igv/assets/139537326/eeef616b-be9b-4c75-b540-185718288857)
![スクリーンショット 2024-02-01 14 35 53](https://github.com/igvteam/igv/assets/139537326/1b0eca62-e132-4fa3-9540-91fa335eb03d)
![スクリーンショット 2024-02-01 14 32 05](https://github.com/igvteam/igv/assets/139537326/05e13f8d-97b1-488f-aed0-0040dbae0eba)
When I was preparing these screenshots, I found two points.
1. When zooming out, IGV seems to keep higher value to sum/average local area, which is good for most of case, but may not for DNA methylation (at CG site).
2. I still need to discriminate 5mC/5hmC level at CG and others and this is true to most of researchers on DNA methylation study.
I hope this is useful comment to IGV.

**Comment by eesiribloom** [comment]
Yes a CGI is a CpG Island. But I think if there was a way to toggle to see 5mC at CG dinucloetides this would be most helpful. Right now if you call 5mC in ONT data and visualise modified bases on IGV, it's not helpful as the plot becomes overwhelming. It's usually only informative to visualise 5mC at CG dinucleotides. It's pretty easy to generate a bed file of [CG dinucleotides from a reference](https://nanoporetech.github.io/modkit/intro_motif_bed.html) or download tracks of CpG Islands from [UCSC ](https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=578954849_wF1QP81SIHdfr8b0kmZUOcsZcHYr&clade=mammal&org=Human&db=hg38&hgta_group=regulation&hgta_track=knownGene&hgta_table=0&hgta_regionType=genome&position=chr9%3A133252000-133280861&hgta_outputType=primaryTable&hgta_outFileName=). If there was a way to restrict modified bases to these regions which could be supplied by the user in a bed file for the relevant reference genome ? Just brainstorming here...

**Comment by jrobinso** [comment]
No there is no way currently to do that.   If you really don't care about 5mC elsewhere wouldn't be easier to just remove the calls at other locations in the BAM file?   

WRT implementations, it might be easier to consider only CG dinculeotides in the read sequence, and more consistent with base calling which is not reference dependent.   That would be consistent with the "--motif" option of the tool you mention above.

**Comment by jrobinso** [comment]
I was looking into this and my test data, from both Pacbio and ONT,  has 5mC calls only at CG sites,  there appears to be no 5mC calls at any other site.    If either of you can share or point to a public dataset (bam file) with 5mC calls at non-CG sites I will look into this further.

**Comment by jrobinso** [comment]
Thanks for the link.  It would be really helpful if the data had some negative strand reads,  which is important in this application, but the simulated reads are all aligned to the + strand.    If you find or can provide a small dataset of real data (as opposed to simulated) that would be helpful here.  But this is a start.

**Comment by jrobinso** [comment]
I've added a preference to set a context for all "C" modifications to the base mods tab.   The context filter is based on the read, not the reference, so a "CG" in the read will match a CG context irrespective of the reference at that location.
